lib/propolis/src/block/mod.rs

```diff
@@ -650,3 +650,260 @@ impl DriverInner {
+
+    /// Semaphore to activate idle worker to pull request from the queue
+    worker_activate: Semaphore,
+
+    /// Number of requests being actively serviced
+    active_req_count: AtomicUsize,
+}
+
+impl Inner {
+    fn new() -> Arc<Self> {
+        Arc::new(Self {
+            bdev: OnceCell::new(),
+            acc_mem: OnceCell::new(),
+            sched_ctrl: Mutex::new(None),
+
+            queue: TokioMutex::new(VecDeque::new()),
+
+            wake: Arc::new(Notify::new()),
+            worker_idle: Semaphore::new(0),
+            worker_activate: Semaphore::new(0),
+
+            active_req_count: AtomicUsize::new(0),
+        })
+    }
+
+    fn attach(self: &Arc<Self>, bdev: Arc<dyn Device>) {
+        let _old_bdev = self.bdev.set(bdev.clone());
+        assert!(_old_bdev.is_ok(), "driver already attached");
+        let _old_acc = self.acc_mem.set(bdev.accessor_mem());
+        assert!(_old_acc.is_ok(), "driver already attached");
+
+        let sched_wake = self.wake.clone();
+        let (mut thdl, tctrl) =
+            TaskHdl::new_held(Some(Box::new(move || sched_wake.notify_one())));
+        let sched_self = Arc::clone(self);
+        tokio::spawn(async move {
+            sched_self.schedule_requests(&mut thdl).await;
+        });
+        let _old_ctrl = self.sched_ctrl.lock().unwrap().replace(tctrl);
+        assert!(_old_ctrl.is_none(), "driver already attached");
+
+        let wake_self = Arc::clone(self);
+        bdev.set_notifier(Some(Box::new(move |_bdev| {
+            wake_self.wake.notify_one()
+        })));
+    }
+
+    fn with_ctrl(&self, f: impl FnOnce(&mut TaskCtrl)) {
+        f(self.sched_ctrl.lock().unwrap().as_mut().expect("scheduler started"))
+    }
+
+    fn start(&self) {
+        self.with_ctrl(|ctrl| {
+            let _ = ctrl.run();
+        });
+    }
+    fn pause(&self) {
+        self.with_ctrl(|ctrl| {
+            let _ = ctrl.hold();
+        });
+    }
+    fn resume(&self) {
+        self.with_ctrl(|ctrl| {
+            let _ = ctrl.run();
+        });
+    }
+    fn halt(&self) {
+        self.with_ctrl(|ctrl| {
+            let _ = ctrl.exit();
+        });
+    }
+
+    fn worker(self: &Arc<Self>) -> WorkerHdl {
+        WorkerHdl::new(self.clone())
+    }
+
+    /// Attempt to grab a request from the block device (if one's available)
+    async fn next_req(&self) -> Request {
+        let bdev = self.bdev.get().expect("attached block device");
+        loop {
+            if let Some(req) = bdev.next() {
+                return req;
+            }
+
+            // Don't busy-loop on the device but just wait for it to wake us
+            // when the next request is available
+            self.wake.notified().await;
+        }
+    }
+
+    fn complete_req(&self, req: Request, res: Result) {
+        let bdev = self.bdev.get().expect("attached bdev");
+        req.complete(res, bdev.as_ref());
+        match self.active_req_count.fetch_sub(1, Ordering::SeqCst) {
```

Comment:
This (and the other use of `SeqCst` in `schedule_requests`) can be `AcqRel` right? `active_req_count` is the only atomic in question here so there's no other location we're trying to synchronize with such as to require the consistent total ordering `SeqCst` guarantees.

Reply:
Done